docs/vm-presets.md

```diff
@@ -0,0 +1,289 @@
+Virtual Machine Presets
+=============================
+
+`VirtualMachinePresets` are an extension to general `VirtualMachine`
+configuration behaving much like `PodPresets` from Kubernetes. When a
+`VirtualMachine` is created, any applicable `VirtualMachinePresets`
+will be applied to the existing spec for the `VirtualMachine`. This allows
+for re-use of common settings that should apply to multiple `VirtualMachines`.
+
+
+Implementation
+------------------------
+
+`VirtualMachinePresets` are implemented as a Kubernetes `Initializer`. This
+means the `VirtualMachine` resource is modified before it is visible to or
+processed by any other component of KubeVirt.
+
+Once a `VirtualMachinePreset` is successfully applied to a `VirtualMachine`,
+the `VirtualMachine` will be marked with an annotation to indicate that it
+was applied. If an error occurs while a `VirtualMachinePreset` is being applied
+(for example, if a conflict occurs), none of the `VirtualMachinePreset` will be
+applied.
+
+
+Usage
+------------------------
+
+KubeVirt uses Kubernetes `Labels` and `Selectors` to determine which
+`VirtualMachinePresets` apply to any given `VirtualMachine`, similarly to how
+`PodPresets` work in Kubernetes. The `VirtualMachine` is marked with an
+Annotation upon successful completion.
+
+Any domain structure can be listed in the `spec` of a `VirtualMachinePreset`.
+e.g. Clock, Features, Memory, CPU, or Devices such network interfaces or disks.
+All elements of the `spec` section of a `VirtualMachinePreset` will be applied
+to the `VirtualMachine`.
+
+
+Conflicts
+------------------------
+
+`VirtualMachinePresets` use a similar conflict resolution strategy to
+Kubernetes `PodPresets`. If a portion of the domain spec is present in both a
+`VirtualMachine` and a `VirtualMachinePreset` and both resources have the
+identical information, then no error will occur and `VirtualMachine` creation
+will continue normally. If however there is a conflict between the resources,
+an error will occur and the `VirtualMachine` will not be created. For example:
+If both the `VirtualMachine` and `VirtualMachinePreset` define a `Volume`, but
+use different paths for the same name, KubeVirt will note the conflict.
```

Comment:
This makes sense to me.

Looking at some of this in more detail, some of these merge conflicts can get kind of complicated.  We talked briefly about trying to figure out if there was a way to generate code to automate this merging. I'm coming to the same conclusion you are now.  I don't think we can automate any of this. Merging looks like it is going to require us to take each field on a case by case basis. 

For example, two separate disks might reference different volumes but attempt to point to the same dev (like dev=vda). That should probably be a conflict and its not just generically tied to the name of the disk or volume where we could do a deep equals.

Reply:
It might still be possible to generate code based on annotation which we put on fields on the vm. To make some up: `presetWholeStructOnly`, `presetPerStructMember`, ...

We don't need it right now.

Reply:
I believe we need to be really careful here. Presets may contain anything, and initially GUI will depend on having the VM as complete as possible, but if it contains a volume then indeed there's going to be a conflict right away.
There should be some guidelines written down what a preset should and should not contain. Non-binding, but we do need some expectations/assumptions to be clear.